ouroboros-network/src/Ouroboros/Network/Mux.hs

```diff
@@ -0,0 +1,341 @@
+
+data MiniProtocolDescription m = MiniProtocolDescription {
+      mpdId        :: MiniProtocolId
+    , mpdInitiator :: Duplex m m CBOR.Encoding BS.ByteString -> m ()
+    , mpdResponder :: Duplex m m CBOR.Encoding BS.ByteString -> m ()
+    }
+
+newtype MiniProtocolDescriptions m = MiniProtocolDescriptions (M.Map MiniProtocolId (MiniProtocolDescription m))
+
+newtype MiniProtocolDispatch m = MiniProtocolDispatch (M.Map (MiniProtocolId, MiniProtocolMode)
+                                                             (TBQueue m BL.ByteString))
+
+data MiniProtocolMode = ModeInitiator | ModeResponder deriving (Eq, Ord, Show)
+
+negMiniProtocolMode :: MiniProtocolMode -> MiniProtocolMode
+negMiniProtocolMode ModeInitiator = ModeResponder
+negMiniProtocolMode ModeResponder = ModeInitiator
+
+data MuxSDU = MuxSDU {
+      msTimestamp :: !RemoteClockModel
+    , msId        :: !MiniProtocolId
+    , msMode      :: !MiniProtocolMode
+    , msLength    :: !Word16
+    , msBlob      :: !BL.ByteString
+    }
+
+encodeMuxSDU :: MuxSDU -> BL.ByteString
+encodeMuxSDU sdu =
+  let hdr = Bin.runPut enc in
+  BL.append hdr $ msBlob sdu
+  where
+    enc = do
+        Bin.putWord32be $ unRemoteClockModel $ msTimestamp sdu
+        putId (msId sdu) (putMode $ msMode sdu)
+        Bin.putWord16be $ fromIntegral $ BL.length $ msBlob sdu
+
+    putId Muxcontrol mode    = Bin.putWord16be $ 0 .|. mode
+    putId DeltaQ mode        = Bin.putWord16be $ 1 .|. mode
+    putId ChainSync mode     = Bin.putWord16be $ 2 .|. mode
+    putId Blockdownload mode = Bin.putWord16be $ 3 .|. mode
+    putId TxSubmission mode  = Bin.putWord16be $ 4 .|. mode
+
+    putMode :: MiniProtocolMode -> Word16
+    putMode ModeInitiator = 0
+    putMode ModeResponder = 0x8000
+
+decodeMuxSDUHeader :: BL.ByteString -> Maybe MuxSDU
+decodeMuxSDUHeader buf =
+    case Bin.runGetOrFail dec buf of
+         Left  (_, _, _)  -> Nothing
+         Right (_, _, ph) -> Just ph
+
+  where
+    dec = do
+        ts <- Bin.getWord32be
+        mid <- Bin.getWord16be
+        len <- Bin.getWord16be
+        return $ MuxSDU (RemoteClockModel ts) (getId (mid .&. 0x7fff)) (getMode (mid .&. 0x8000))
+                        len BL.empty
+
+    getMode 0      = ModeInitiator
+    getMode 0x8000 = ModeResponder
+    getMode _      = error "impossible use of bitmask" -- XXX
+
+    getId 0 = Muxcontrol
+    getId 1 = DeltaQ
+    getId 2 = ChainSync
+    getId 3 = Blockdownload
+    getId 4 = TxSubmission
+    getId a = error $ "unknow miniprotocol " ++ show a -- XXX
+
+class MuxBearer m where
+  type LocalClockModel m :: *
+  type AssociationDetails m :: *
+  type MuxBearerHandle m :: *
+  type ResponderHandle m :: *
+  initiator :: MiniProtocolDescriptions m -> AssociationDetails m -> AssociationDetails m -> m ()
+  responder :: MiniProtocolDescriptions m -> AssociationDetails m -> m (ResponderHandle m)
+  killResponder :: ResponderHandle m -> m ()
+  sduSize :: MuxBearerHandle m-> m Word16
```

Comment:
It is possible that it could change due to routing updates or network configuration changes. It is not strictly needed for TCP but would be very useful if we implement support for datagram protocols in the future.